src/compas_timber/design/workflow.py

```diff
@@ -34,6 +37,87 @@ def comply(self, beams):
+                JointTopology.TOPO_X: TopologyRule(JointTopology.TOPO_X, XHalfLapJoint),
+            }
+        for rule in rules:  # separate category and topo and direct joint rules
+            if rule.__class__.__name__ == "TopologyRule":
+                topo_rules[rule.topology_type] = TopologyRule(
+                    rule.topology_type, rule.joint_type, **rule.kwargs
+                )  # overwrites, meaning last rule wins
+        return [rule for rule in topo_rules.values() if rule is not None]
+
+    @staticmethod
+    def joints_from_beams_and_rules(beams, rules, max_distance=1e-6):
+        """Culls direct rules that are not applicable to the given beams.
+
+        Parameters
+        ----------
+        beams : list(:class:`~compas_timber.parts.Beam`)
+            A list of beams to be joined.
+        rules : list(:class:`~compas_timber.design.JointRule`)
+            A list of joint rules to be applied.
+        max_distance : float, optional
+            The maximum distance to consider two beams as intersecting.
+
+        Returns
+        -------
+        list(:class:`~compas_timber.design.JointDefinition`)
+            A list of joint definitions that can be applied to the given beams.
+
+        """
+        beams = beams if isinstance(beams, list) else list(beams)
+        direct_rules = JointRule.get_direct_rules(rules)
+        solver = ConnectionSolver()
+
+        beam_pairs = solver.find_intersecting_pairs(beams, rtree=True, max_distance=max_distance)
+        joint_defs = []
+        for rule in direct_rules:
+            joint_defs.append(JointDefinition(rule.joint_type, rule.beams, **rule.kwargs))
+
+        while beam_pairs:
+            pair = beam_pairs.pop()
+            match_found = False
+            for rule in direct_rules:  # see if pair is used in a direct rule
+                if rule.comply(pair):
+                    match_found = True
+                    break
+
+            if not match_found:
+                for rule in JointRule.get_category_rules(rules):  # see if pair is used in a category rule
+                    if rule.comply(pair):
+                        match_found = True
+                        joint_defs.append(JointDefinition(rule.joint_type, rule.reorder(pair), **rule.kwargs))
+                        break
+
+            if not match_found:
+                for rule in JointRule.get_topology_rules(rules):  # see if pair is used in a topology rule
+                    comply, pair = rule.comply(pair)
+                    if comply:
+                        match_found = True
+                        joint_defs.append(JointDefinition(rule.joint_type, pair, **rule.kwargs))
+                        break
+            if not match_found:
+                print(
+                    "Beam pairs could not be resolved by the rules: ",
+                    "({}, {})".format(list(pair)[0].key, list(pair)[1].key),
+                )  # TODO: add something to catch unresolved pairs
```

Comment:
since this is inside the library now, prints and such (feedback to user should be handled by the calling code that's user facing, AKA in the GH component)

Reply:
What about raise warning()?